File: proyecto2/harvest-scheduling/src/main/scala/HarvestScheduling.scala
import scala.util.Random;

class HarvestScheduling(initialSolution: Array[Array[ForestUnit]], s: Int) extends Runnable {

  private val plan = initialSolution
  private val seed = s
  private val rng = new Random(seed)

  def run() {
    val startTime = System.nanoTime
    val dec = Array.ofDim[Int](Parameters.Periods, Parameters.Units)

    /* Asignar las variables de decisión para que cumplan la restricción de
       singularidad */
    for(i <- 0 until Parameters.Units) {
      val pos = rng.nextInt(Parameters.Periods)
      val prob = rng.nextFloat
      if(prob < 0.5) dec(pos)(i) = 1
    }

    val schedule = new SchedulePlan(plan, dec)
    val objective = schedule.objective
    println(f"Inicial ($seed%d): $objective%.2f")

    val bestSolution = TabuSearch.run(Parameters.Iterations, schedule, seed)
    val bestObjective = bestSolution.objective

    println(f"Mejor solución ($seed%d): $bestObjective%.2f")
    val endTime = System.nanoTime
    val totalTime = (endTime - startTime) / 1000000000.0
    println(f"Tiempo total ($seed%d): $totalTime%.2f\n")
  }

}

File: proyecto2/harvest-scheduling/src/main/java/lol/HarvestScheduling.java
package lol;

import java.util.Random;
import java.io.*;
import scala.io.Source;

public class HarvestScheduling implements Runnable {

  private ForestUnit[][] plan;
  private int seed;
  private Random rng;

  public HarvestScheduling(ForestUnit[][] initialSolution, int seed){
    plan = initialSolution;
    this.seed = seed;
    rng = new Random(seed);
  }

  public void run() {

    PrintWriter writer = null, writer2 = null;
    try {
      writer =
        new PrintWriter(new File(String.format("src/etc/solutions/%d", seed)));
      writer2 =
        new PrintWriter(new File(String.format("src/etc/solutions/%d.svg", seed)));
    } catch(FileNotFoundException e){}

    long startTime = System.nanoTime();
    int[][] dec = new int[Parameters.PERIODS][Parameters.UNITS];

    /* Asignar las variables de decisión para que cumplan la restricción de
       singularidad */
    for (int i = 0; i < Parameters.UNITS; i++) {
      int pos = rng.nextInt(Parameters.PERIODS);
      float prob = rng.nextFloat();
      if(prob < 0.5)
        dec[pos][i] = 1;
    }

    SchedulePlan schedule = new SchedulePlan(plan, dec);

    SchedulePlan bestSolution = TabuSearch.run(Parameters.ITERATIONS, schedule, seed);

    long endTime = System.nanoTime();
    double totalTime = (endTime - startTime) / 1000000000.0;

    String l1 = String.format("Solución inicial: %.2f", schedule.objective());
    String l2 = String.format("Mejor solución: %.2f", bestSolution.objective());
    String l3 = String.format("Semilla: %d", seed);
    String l4 = String.format("Tiempo ejecución: %.2f", totalTime);

    writer.println(l1);
    writer.println(l2);
    writer.println(l3);
    writer.println(l4);

    writer2.println(bestSolution.toSVG());

    writer.flush();
    writer.close();
    writer2.flush();
    writer2.close();
  }

}
